File: api/template/template.go
package template

import (
	"github.com/gin-gonic/gin"
	"github.com/nireo/gello/lib/middlewares"
)

// ApplyRoutes to gin engine
func ApplyRoutes(r *gin.RouterGroup) {
	template := r.Group("/template")
	{
		template.GET("/", getTemplates)
		template.GET("/:id", getTemplateWithID)
		template.POST("/", middlewares.Authorized, createTemplate)
		template.POST("/:id", middlewares.Authorized, applyTemplate)
		template.PATCH("/:id", middlewares.Authorized, updateTemplate)
		template.DELETE("/:id", middlewares.Authorized, deleteTemplate)
	}
}

File: client/src/services/template.ts
import axios from 'axios';
import { CreateTemplate } from '../interfaces/Template';
const baseUrl: string = '/api/template';
let token: string | null = null;

export const setToken = (newToken: string) => {
  token = `bearer ${newToken}`;
};

const getConfig = () => ({
  headers: {
    Authorization: token,
  },
});

export const getTemplates = async () => {
  const response = await axios.get(baseUrl, getConfig());
  return response.data;
};

export const getTemplateWithID = async (id: string) => {
  const response = await axios.get(`${baseUrl}/${id}`);
  return response.data;
};

export const createTemplate = async (newTemplate: CreateTemplate) => {
  const response = await axios.post(baseUrl, newTemplate, getConfig());
  return response.data;
};

export const deleteTemplate = async (id: string) => {
  const response = await axios.delete(`${baseUrl}/${id}`, getConfig());
  return response.data;
};

export const updateTemplate = async (
  updatedTemplate: CreateTemplate,
  id: string
) => {
  const response = await axios.patch(
    `${baseUrl}/${id}`,
    { updatedTemplate },
    getConfig()
  );
  return response.data;
};
